File: CS/DetailCollectionEditing/ViewModels/MainViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.Xpf;
using EntityFrameworkIssues.Issues;
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Threading.Tasks;

namespace DetailCollectionEditing {
    public class MainViewModel : ViewModelBase {
        IssuesContext _Context;
        IList<User> _ItemsSource;

        public IList<User> ItemsSource {
            get {
                if(_ItemsSource == null && !IsInDesignMode) {
                    _Context = new IssuesContext();
                    _ItemsSource = _Context.Users.ToList();
                }
                return _ItemsSource;
            }
        }
        [Command]
        public void CreateEditEntityViewModel(CreateEditItemViewModelArgs args) {
            var context = new IssuesContext();
            User item;
            if(args.Key != null)
                item = context.Users.Find(args.Key);
            else {
                item = new User();
                context.Entry(item).State = EntityState.Added;
            }
            args.ViewModel = new EditUserViewModel(item, context, title: (args.IsNewItem ? "New " : "Edit ") + nameof(User));
        }
        [Command]
        public void ValidateRow(EditFormRowValidationArgs args) {
            var context = (IssuesContext)args.EditOperationContext;
            context.SaveChanges();
        }
        [Command]
        public void ValidateRowDeletion(ValidateRowDeletionArgs args) {
            var item = (User)args.Items.Single();
            _Context.Users.Remove(item);
            _Context.SaveChanges();
        }
        [Command]
        public void DataSourceRefresh(DataSourceRefreshArgs args) {
            _ItemsSource = null;
            _Context = null;
            RaisePropertyChanged(nameof(ItemsSource));
        }
    }
}

File: VB/DetailCollectionEditing/ViewModels/MainViewModel.vb
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.Xpf
Imports EntityFrameworkIssues.Issues
Imports System.Collections.Generic
Imports System.Data.Entity
Imports System.Linq

Namespace DetailCollectionEditing

    Public Class MainViewModel
        Inherits ViewModelBase

        Private _Context As IssuesContext

        Private _ItemsSource As IList(Of User)

        Public ReadOnly Property ItemsSource As IList(Of User)
            Get
                If _ItemsSource Is Nothing AndAlso Not IsInDesignMode Then
                    _Context = New IssuesContext()
                    _ItemsSource = _Context.Users.ToList()
                End If

                Return _ItemsSource
            End Get
        End Property

        <Command>
        Public Sub CreateEditEntityViewModel(ByVal args As CreateEditItemViewModelArgs)
            Dim context = New IssuesContext()
            Dim item As User
            If args.Key IsNot Nothing Then
                item = context.Users.Find(args.Key)
            Else
                item = New User()
                context.Entry(item).State = EntityState.Added
            End If

            args.ViewModel = New EditUserViewModel(item, context, title:=If(args.IsNewItem, "New ", "Edit ") & NameOf(User))
        End Sub

        <Command>
        Public Sub ValidateRow(ByVal args As EditFormRowValidationArgs)
            Dim context = CType(args.EditOperationContext, IssuesContext)
            context.SaveChanges()
        End Sub

        <Command>
        Public Sub ValidateRowDeletion(ByVal args As ValidateRowDeletionArgs)
            Dim item = CType(args.Items.[Single](), User)
            _Context.Users.Remove(item)
            _Context.SaveChanges()
        End Sub

        <Command>
        Public Sub DataSourceRefresh(ByVal args As DataSourceRefreshArgs)
            _ItemsSource = Nothing
            _Context = Nothing
            RaisePropertyChanged(NameOf(MainViewModel.ItemsSource))
        End Sub
    End Class
End Namespace
